### DISCUSSION TITLE
Custom build tasmota bin on ESP32 dev board - resetting from watchdog

### DISCUSSION THREAD

**Discussion by Slav3k** [arendst/Tasmota#20618] [Support request] [answered]
Hello guys,

need to make mz HDC1080 sensor work with tasmota. I tried to flash the "vanilla" tasmota bin from the https://tasmota.github.io/install/. That worked perfectly my dev board was able to find my wifi using the web's dialog box.

But ofcourse the vanilla binary is not supporting my sensor HDC1080. Therefore I cloned the repo, made following user%config%override.h:

```
// !!! Remember that your changes GOES AT THE BOTTOM OF THIS FILE right before the last #endif !!!
*/ #ifdef STA_SSID1
undef STA_SSID1
endif
define STA_SSID1 "My wifi's ssid"
ifdef STA_PASS1
undef STA_PASS1
endif
define STA_PASS1 "wifi pass"
ifndef USE_HDC1080
define USE_HDC1080
endif
#endif  // _USER_CONFIG_OVERRIDE_H_
```
Compiled successfully using command:

`platformio run -e tasmota32`
resulting bin I flashed using the web flasher with custom binary.

Unfortunatelly the custom tasmota is not connecting to my wifi. Also I can see following log in the console from the debive:

```
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0x10
clk_drv:0x00,q_drv:0x01,d_drv:0x00,cs0_drv:0x02,hd_drv:0x00,wp_drv:0x04
mode:DOUT, clock div:2
cmd len 52953120
ets_main.c 371 
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0x10
clk_drv:0x00,q_drv:0x01,d_drv:0x00,cs0_drv:0x02,hd_drv:0x00,wp_drv:0x04
mode:DOUT, clock div:2
cmd len 52953120
ets_main.c 371 
ets Jun  8 2016 00:22:57
```

Seems that it is getting constantly restarted from the watchdog. Anybody knows what am I doing wrong?

Many thanks for your support! :)

**Comment by sfromis** [comment, marked as answer]
You probably flashed the wrong binary file, like `tasmota32.bin` instead of `tasmota32.factory.bin` required for initial wired flashing. 

When already having the board flashed with Tasmota, instead of using the web flasher, I'd have recommended the easier approach of doing an upgrade via the Tasmota web UI, where `tasmota32.bin` would be the correct file to use this way.

**Comment by Slav3k** [comment]
Many thanks, that was indeed the issue. 

May I ask a followup question? Now that my custom tasmota build is flashed and the device is reachable via browser, I have configured the I2C pins that are connected to my HDC1080 like this:

<img width="289" alt="image" src="https://github.com/arendst/Tasmota/assets/23500179/8ca5e8e0-c53f-46a4-b0b9-5fcaad290804">

What is the **second number** after the pins? Is that like I2C channel (either I can connect the pins to channel 1 or channel 2)?

Should my HDC1080 be now recognized automatically by tasmota (since I flashed custom build that should support HDC1080)? I think my wiring is correct, the HDC1080 has both GND and 3V3, and the SCL and SDA is connected to pins 21 and 22 of ESP32. 

EDIT: The wiring is correct, when I measure with voltmeter, I can see 3V3 present against ground, and the SDA / SCL are being pulled up to near 3V3.

Many thanks for help with my noob questions :)

**Reply by sfromis** [reply]
The dropdown number is in case of using a driver supporting use of the 2nd I2C bus (most drivers do not). As "channel" is not not part of common I2C terminology, I can't confirm nor deny what you ask about. I2C supports many sensors on the same bus, with separate addresses, and this could also be seen as "channels" in everyday usage.

If you are using a build including `USE_HDC1080`, index 45 should be included in the output of command `I2CDriver` and address `0x40` from `I2CScan`. https://tasmota.github.io/docs/I2CDEVICES/

**Comment by Slav3k** [comment]
Sorry I messed up the terminology - I meant the i2c bus when I said "channel". OK I will keep it on I2C bus 1. 

Now with the HDC1080 I performed this:
```
13:15:47.910 CMD: I2cDriver45 1
13:15:47.917 RSL: RESULT = {"I2CDriver":"7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,45,46,48,58,62,65,69,76,77,82"}
```

Then my tasmota rebooted

then I tried the I2Cscan
```
13:18:59.239 CMD: I2CScan
13:19:00.244 RSL: RESULT = {"I2CScan":"Error 5 at 0x01"}
```

I was expecting to see resul 0x40, but I get error instead. Any ideas?

**Reply by sfromis** [reply]
That error is usually bad wiring. You wrote:
`and the SCL and SDA is connected to pins 21 and 22 of ESP32.`
If those words are to be read very literally, as "respectively", it is not the same order as your config. And yes, this may be over-the-top nitpicking, but computers are very fussy about details.

If you have other active drivers for the same I2C addresss, it could make sense to disable them, like index 9 and 14. Not sure if there would be any conflict, just to be sure.

(The driver was most probably already enabled, unless you specifically disabled it earlier).

FWIW, I do have HDC1080 working on an ESP32, along with a bunch of other sensors.

**Comment by Slav3k** [comment]
The wires were too long. I tried to read the sensor using simple arduino program and it was also not working. Then I shortened the wires and voila :) 

Many thanks for your support! 

Now comes the last step of my experiment - learn how to publish the readings via MQTT and integrate this to my automation center (mycodo running on raspberry pi).

**Reply by sfromis** [reply]
By default, Tasmota is automatically publishing the readings every `TelePeriod`, default 300 seconds. It should be simple to subscribe to the payloads, and any decent automation should be able to parse JSON data painlessly.

And yeah, the I2C bus is designed for connections in a device, not for longer wiring away from the box.

**Reply by Slav3k** [reply]
Nice! Is it that the `TelePeriod` is also the "sampling frequency"? By sampling frequency I mean the frequency of acquiring measurements from sensors that are being sent as telemetry?

**Reply by sfromis** [reply]
No, the sampling frequency is much higher, about every second. And the web UI front page (when active) gets updated at intervals of 2.345 seconds by default.
